crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,324 @@
+    dependency_log_levels: Option<String>,
+    client_info: Option<&ClientInfo>,
+) {
+    let log_level = resolve_log_level(log_level);
+    let dependency_log_levels = resolve_dependency_log_levels(dependency_log_levels);
+
+    let writer = if client_info.is_some_and(|client_info| {
+        client_info.name.starts_with("Zed") || client_info.name.starts_with("Visual Studio Code")
+    }) {
+        // These IDEs are known to support `window/logMessage` well
+        BoxMakeWriter::new(LogWriterMaker::new(log_tx))
+    } else if is_test_client(client_info) {
+        // Ensures a standard `cargo test` captures output unless `-- --nocapture` is used
+        BoxMakeWriter::new(TestWriter::default())
+    } else {
+        // Fallback for other editors / IDEs
+        BoxMakeWriter::new(std::io::stderr)
+    };
+
+    let layer = tracing_subscriber::fmt::layer()
+        // Disable ANSI escapes, those are not supported in Code
+        .with_ansi(false)
+        // Display source code file paths
+        .with_file(true)
+        // Display source code line numbers
+        .with_line_number(true)
+        // Don't display the thread ID or thread name
+        .with_thread_ids(false)
+        .with_thread_names(false)
+        // Don't display the event's target (module path).
+        // Mostly redundant with file paths.
+        .with_target(false)
+        // Display local time rather than UTC
+        .with_timer(LocalTime::rfc_3339())
+        // Display the log level
+        .with_level(true)
+        .with_writer(writer)
+        .with_filter(log_filter(log_level, dependency_log_levels));
+
+    let subscriber = tracing_subscriber::Registry::default().with(layer);
+
+    if is_test_client(client_info) {
+        // During parallel testing, `set_global_default()` gets called multiple times
+        // per process. That causes it to error, but we ignore this.
+        tracing::subscriber::set_global_default(subscriber).ok();
+    } else {
+        tracing::subscriber::set_global_default(subscriber)
+            .expect("Should be able to set the global subscriber exactly once.");
+    }
+
+    tracing::info!("Logging initialized with level: {log_level}");
+}
+
+/// We use a special `TestWriter` during tests to be compatible with `cargo test`'s
+/// typical output capturing behavior.
+///
+/// Important notes:
+/// - `cargo test` swallows all logs unless you use `-- --nocapture`.
+/// - Tests run in parallel, so logs can be interleaved unless you run `--test-threads 1`.
+///
+/// We use `cargo test -- --nocapture --test-threads 1` on CI because of all of this.
+fn is_test_client(client_info: Option<&ClientInfo>) -> bool {
+    client_info.map_or(false, |client_info| client_info.name == "AirTestClient")
+}
+
+// TODO: Is there a way to generate this at compile time?
+const TARGETS: &[&str] = &[
+    "air_r_factory",
+    "air_r_formatter",
+    "air_r_parser",
+    "air_r_syntax",
+    "fs",
+    "line_ending",
+    "lsp",
+];
```

Comment:
Something rather interesting is that even though we target these crate names, really this ends up being a _prefix match_ on the log `target` field. That means our `lsp_test` crate matches and we got lots of logs around `codec.rs`, so I just commented those out for now.

---

We could actually take advantage of the prefix matching if we wanted to. By prefixing all of our crates with `air_`, this would simplify the code down to a single:

```
filter.with_target("air_", log_level)
```

Which I like because:
- We don't have to worry about updating this list
- We worry less about prefix matching a crate that isn't us